Can't run tokio-fetch example

```
error[E0282]: type annotations needed
  --> src/fetch.rs:26:10
   |
26 |         .downcast_or_throw::<JsObject, _>(cx)?;
   |          ^^^^^^^^^^^^^^^^^ cannot infer type
   |
   = note: type must be known at this point

error[E0599]: no method named `downcast_or_throw` found for struct `neon::handle::Handle<'_, _>` in the current scope
  --> src/fetch.rs:26:10
   |
26 |         .downcast_or_throw::<JsObject, _>(cx)?;
   |          ^^^^^^^^^^^^^^^^^ method not found in `neon::handle::Handle<'_, _>`
   |
   = note: the method was found for
           - `neon::handle::Handle<'a, T>`

Some errors have detailed explanations: E0282, E0599.
For more information about an error, try `rustc --explain E0282`.
```

--- a/examples/async-sqlite/Cargo.toml
+++ b/examples/async-sqlite/Cargo.toml
@@ -13,6 +13,6 @@ crate-type = ["cdylib"]
 rusqlite = "0.25"
 
 [dependencies.neon]
-version = "0.9"
+version = "0.10.1"
 default-features = false
-features = ["napi-6", "event-queue-api", "try-catch-api"]
+features = ["napi-6", "channel-api", "promise-api", "try-catch-api"]

--- a/examples/async-sqlite/index.js
+++ b/examples/async-sqlite/index.js
@@ -4,10 +4,6 @@ const { promisify } = require("util");
 
 const { databaseNew, databaseClose, databaseInsert, databaseGetById } = require("./index.node");
 
-// Convert the DB methods from using callbacks to returning promises
-const databaseInsertAsync = promisify(databaseInsert);
-const databaseGetByIdAsync = promisify(databaseGetById);
-
 // Wrapper class for the boxed `Database` for idiomatic JavaScript usage
 class Database {
     constructor() {
@@ -18,11 +14,11 @@ class Database {
     // This could be node in several other ways, for example binding assignment
     // in the constructor
     insert(name) {
-        return databaseInsertAsync.call(this.db, name);
+        return databaseInsert.call(this.db, name);
     }
 
     byId(id) {
-        return databaseGetByIdAsync.call(this.db, id);
+        return databaseGetById.call(this.db, id);
     }
 
     close() {

--- a/examples/async-sqlite/package.json
+++ b/examples/async-sqlite/package.json
@@ -11,7 +11,7 @@
   "license": "MIT",
   "devDependencies": {
     "cargo-cp-artifact": "^0.1",
-    "mocha": "^9.1.4"
+    "mocha": "^10.0.0"
   },
   "repository": {
     "type": "git",

--- a/examples/async-sqlite/src/lib.rs
+++ b/examples/async-sqlite/src/lib.rs
@@ -1,10 +1,10 @@
 use std::sync::mpsc;
 use std::thread;
 
-use neon::prelude::*;
+use neon::{prelude::*, types::Deferred};
 use rusqlite::Connection;
 
-type DbCallback = Box<dyn FnOnce(&mut Connection, &Channel) + Send>;
+type DbCallback = Box<dyn FnOnce(&mut Connection, &Channel, Deferred) + Send>;
 
 // Wraps a SQLite connection a channel, allowing concurrent access
 struct Database {
@@ -13,8 +13,10 @@ struct Database {
 
 // Messages sent on the database channel
 enum DbMessage {
-    // Callback to be executed
-    Callback(DbCallback),
+    // Promise to resolve and callback to be executed
+    // Deferred is threaded through the message instead of moved to the closure so that it
+    // can be manually rejected.
+    Callback(Deferred, DbCallback),
     // Indicates that the thread should be stopped and connection closed
     Close,
 }
@@ -69,11 +71,11 @@ impl Database {
             // the thread.
             while let Ok(message) = rx.recv() {
                 match message {
-                    DbMessage::Callback(f) => {
+                    DbMessage::Callback(deferred, f) => {
                         // The connection and channel are owned by the thread, but _lent_ to
                         // the callback. The callback has exclusive access to the connection
                         // for the duration of the callback.
-                        f(&mut conn, &channel);
+                        f(&mut conn, &channel, deferred);
                     }
                     // Immediately close the connection, even if there are pending messages
                     DbMessage::Close => break,
@@ -92,9 +94,11 @@ impl Database {
 
     fn send(
         &self,
-        callback: impl FnOnce(&mut Connection, &Channel) + Send + 'static,
+        deferred: Deferred,
+        callback: impl FnOnce(&mut Connection, &Channel, Deferred) + Send + 'static,
     ) -> Result<(), mpsc::SendError<DbMessage>> {
-        self.tx.send(DbMessage::Callback(Box::new(callback)))
+        self.tx
+            .send(DbMessage::Callback(deferred, Box::new(callback)))
     }
 }
 
@@ -125,18 +129,16 @@ impl Database {
     }
 
     // Inserts a `name` into the database
-    // Accepts a `name` and a `callback` as parameters
-    fn js_insert(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    // Accepts a `name` and returns a `Promise`
+    fn js_insert(mut cx: FunctionContext) -> JsResult<JsPromise> {
         // Get the first argument as a `JsString` and convert to a Rust `String`
         let name = cx.argument::<JsString>(0)?.value(&mut cx);
 
-        // Get the second argument as a `JsFunction`
-        let callback = cx.argument::<JsFunction>(1)?.root(&mut cx);
-
         // Get the `this` value as a `JsBox<Database>`
         let db = cx.this().downcast_or_throw::<JsBox<Database>, _>(&mut cx)?;
+        let (deferred, promise) = cx.promise();
 
-        db.send(move |conn, channel| {
+        db.send(deferred, move |conn, channel, deferred| {
             let result = conn
                 .execute(
                     "INSERT INTO person (name) VALUES (?)",
@@ -144,68 +146,72 @@ impl Database {
                 )
                 .map(|_| conn.last_insert_rowid());
 
-            channel.send(move |mut cx| {
-                let callback = callback.into_inner(&mut cx);
-                let this = cx.undefined();
-                let args: Vec<Handle<JsValue>> = match result {
-                    Ok(id) => vec![cx.null().upcast(), cx.number(id as f64).upcast()],
-                    Err(err) => vec![cx.error(err.to_string())?.upcast()],
-                };
+            deferred.settle_with(channel, move |mut cx| {
+                let id = result.or_else(|err| cx.throw_error(err.to_string()))?;
 
-                callback.call(&mut cx, this, args)?;
-
-                Ok(())
+                Ok(cx.number(id as f64))
             });
         })
-        .or_else(|err| cx.throw_error(err.to_string()))?;
+        .into_rejection(&mut cx)?;
 
         // This function does not have a return value
-        Ok(cx.undefined())
+        Ok(promise)
     }
 
     // Get a `name` by `id` value
     // Accepts an `id` and callback as parameters
-    fn js_get_by_id(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    fn js_get_by_id(mut cx: FunctionContext) -> JsResult<JsPromise> {
         // Get the first argument as a `JsNumber` and convert to an `f64`
         let id = cx.argument::<JsNumber>(0)?.value(&mut cx);
 
-        // Get the second argument as a `JsFunction`
-        let callback = cx.argument::<JsFunction>(1)?.root(&mut cx);
-
         // Get the `this` value as a `JsBox<Database>`
         let db = cx.this().downcast_or_throw::<JsBox<Database>, _>(&mut cx)?;
+        let (deferred, promise) = cx.promise();
 
-        db.send(move |conn, channel| {
+        db.send(deferred, move |conn, channel, deferred| {
             let result: Result<String, _> = conn
                 .prepare("SELECT name FROM person WHERE id = ?")
                 .and_then(|mut stmt| stmt.query_row(rusqlite::params![id], |row| row.get(0)));
 
-            channel.send(move |mut cx| {
-                let callback = callback.into_inner(&mut cx);
-                let this = cx.undefined();
-                let args: Vec<Handle<JsValue>> = match result {
-                    // Convert the name to a `JsString` on success and upcast to a `JsValue`
-                    Ok(name) => vec![cx.null().upcast(), cx.string(name).upcast()],
+            deferred.settle_with(channel, move |mut cx| -> JsResult<JsValue> {
+                // If the row was not found, return `undefined` as a success instead
+                // of throwing an exception
+                if matches!(result, Err(rusqlite::Error::QueryReturnedNoRows)) {
+                    return Ok(cx.undefined().upcast());
+                }
 
-                    // If the row was not found, return `undefined` as a success instead
-                    // of throwing an exception
-                    Err(rusqlite::Error::QueryReturnedNoRows) => {
-                        vec![cx.null().upcast(), cx.undefined().upcast()]
-                    }
+                let name = result.or_else(|err| cx.throw_error(err.to_string()))?;
 
-                    // Convert the error to a JavaScript exception on failure
-                    Err(err) => vec![cx.error(err.to_string())?.upcast()],
-                };
-
-                callback.call(&mut cx, this, args)?;
-
-                Ok(())
+                Ok(cx.string(name).upcast())
             });
         })
-        .or_else(|err| cx.throw_error(err.to_string()))?;
+        .into_rejection(&mut cx)?;
 
         // This function does not have a return value
-        Ok(cx.undefined())
+        Ok(promise)
+    }
+}
+
+trait SendResultExt {
+    // Sending a query closure to execute may fail if the channel has been closed.
+    // This method converts the failure into a promise rejection.
+    fn into_rejection<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<()>;
+}
+
+impl SendResultExt for Result<(), mpsc::SendError<DbMessage>> {
+    fn into_rejection<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<()> {
+        self.or_else(|err| {
+            let msg = err.to_string();
+
+            match err.0 {
+                DbMessage::Callback(deferred, _) => {
+                    let err = cx.error(msg)?;
+                    deferred.reject(cx, err);
+                    Ok(())
+                }
+                DbMessage::Close => cx.throw_error("Expected DbMessage::Callback"),
+            }
+        })
     }
 }
 

--- a/examples/async-sqlite/test/db.test.js
+++ b/examples/async-sqlite/test/db.test.js
@@ -29,6 +29,6 @@ describe("SQLite Database", () => {
 
         db.close();
 
-        await assert.rejects(async () => db.byId(5));
+        await assert.rejects(() => db.byId(5), /closed channel/);
     });
 });

--- a/examples/cpu-count/Cargo.toml
+++ b/examples/cpu-count/Cargo.toml
@@ -13,6 +13,6 @@ crate-type = ["cdylib"]
 num_cpus = "1"
 
 [dependencies.neon]
-version = "0.9"
+version = "0.10.1"
 default-features = false
 features = ["napi-6"]

--- a/examples/gzip-stream/Cargo.toml
+++ b/examples/gzip-stream/Cargo.toml
@@ -13,6 +13,6 @@ crate-type = ["cdylib"]
 flate2 = "1"
 
 [dependencies.neon]
-version = "0.10.0-alpha.3"
+version = "0.10.1"
 default-features = false
 features = ["napi-6", "promise-api", "task-api"]

--- a/examples/hello-world/Cargo.toml
+++ b/examples/hello-world/Cargo.toml
@@ -10,6 +10,6 @@ exclude = ["index.node"]
 crate-type = ["cdylib"]
 
 [dependencies.neon]
-version = "0.9"
+version = "0.10.1"
 default-features = false
 features = ["napi-6"]

--- a/examples/tokio-fetch/Cargo.toml
+++ b/examples/tokio-fetch/Cargo.toml
@@ -16,6 +16,6 @@ tokio = { version = "1", features = ["rt-multi-thread"] }
 serde = { version = "1", features = ["derive"] }
 
 [dependencies.neon]
-version = "0.10.0-alpha.3"
+version = "0.10.1"
 default-features = false
 features = ["channel-api", "napi-6", "promise-api"]

--- a/examples/tokio-fetch/src/lib.rs
+++ b/examples/tokio-fetch/src/lib.rs
@@ -20,15 +20,8 @@ fn runtime<'a, C: Context<'a>>(cx: &mut C) -> NeonResult<&'static Runtime> {
 // Get the verson of the currently running node process from [`process.version`](https://nodejs.org/api/process.html#processversion)
 fn node_version<'a, C: Context<'a>>(cx: &mut C) -> NeonResult<String> {
     let global = cx.global();
-
-    let process = global
-        .get(cx, "process")?
-        .downcast_or_throw::<JsObject, _>(cx)?;
-
-    let version = process
-        .get(cx, "version")?
-        .downcast_or_throw::<JsString, _>(cx)?
-        .value(cx);
+    let process = global.get::<JsObject, _, _>(cx, "process")?;
+    let version = process.get::<JsString, _, _>(cx, "version")?.value(cx);
 
     Ok(version)
 }

